worktree.go

```diff
@@ -270,65 +273,99 @@ func (w *Worktree) Reset(opts *ResetOptions) error {
 		}
 	}
 
-	changes, err := w.diffCommitWithStaging(opts.Commit, true)
-	if err != nil {
+	if err := w.setHEADCommit(opts.Commit); err != nil {
 		return err
 	}
 
-	idx, err := w.r.Storer.Index()
-	if err != nil {
-		return err
+	if opts.Mode == SoftReset {
+		return nil
 	}
 
 	t, err := w.getTreeFromCommitHash(opts.Commit)
 	if err != nil {
 		return err
 	}
 
-	for _, ch := range changes {
-		if err := w.checkoutChange(ch, t, idx); err != nil {
+	if opts.Mode == MixedReset || opts.Mode == MergeReset || opts.Mode == HardReset {
+		if err := w.resetIndex(t); err != nil {
 			return err
 		}
 	}
 
-	if err := w.r.Storer.SetIndex(idx); err != nil {
-		return err
+	if opts.Mode == MergeReset || opts.Mode == HardReset {
+		if err := w.resetWorktree(t); err != nil {
+			return err
+		}
 	}
 
-	return w.setHEADCommit(opts.Commit)
+	return nil
 }
 
-func (w *Worktree) containsUnstagedChanges() (bool, error) {
-	ch, err := w.diffStagingWithWorktree()
+func (w *Worktree) resetIndex(t *object.Tree) error {
+	idx, err := w.r.Storer.Index()
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	return len(ch) != 0, nil
-}
-
-func (w *Worktree) setHEADCommit(commit plumbing.Hash) error {
-	head, err := w.r.Reference(plumbing.HEAD, false)
+	changes, err := w.diffTreeWithStaging(t, true)
 	if err != nil {
 		return err
 	}
 
-	if head.Type() == plumbing.HashReference {
-		head = plumbing.NewHashReference(plumbing.HEAD, commit)
-		return w.r.Storer.SetReference(head)
+	for _, ch := range changes {
+		a, err := ch.Action()
+		if err != nil {
+			return err
+		}
+
+		var name string
+		var e *object.TreeEntry
+
+		switch a {
+		case merkletrie.Modify, merkletrie.Insert:
+			name = ch.To.String()
+			e, err = t.FindEntry(name)
+			if err != nil {
+				return err
+			}
+		case merkletrie.Delete:
+			name = ch.From.String()
+		}
+
+		_, _ = idx.Remove(name)
```

Comment:
can this operation fail? If so, should we leave it unhandled?

Reply:
the only error returned by Remove, is `ErrEntryNotFound` and we don't care about it.